backend/battle/forms.py

```diff
@@ -20,13 +27,43 @@ def __init__(self, *args, **kwargs):
         super(BattleForm, self).__init__(*args, **kwargs)
         self.fields['creator'].widget = forms.HiddenInput()
 
+    def clean_opponent(self):
+        opponent_email = self.cleaned_data["opponent"]
+        try:
+            opponent = User.objects.get(email=opponent_email)
+        except User.DoesNotExist:
+            self.is_guest = True
+            opponent = User.objects.create(email=opponent_email)
+            random_password = get_random_string(length=64)
+            opponent.set_password(random_password)
+            opponent.save()
+        return opponent
+
     def clean(self):
         cleaned_data = super().clean()
         if 'creator' not in cleaned_data:
             raise forms.ValidationError("ERROR: You need to be logged.")
 
         if cleaned_data['opponent'] == cleaned_data['creator']:
             raise forms.ValidationError("ERROR: You can't challenge yourself.")
+        return cleaned_data
+
+    def save(self, commit=True):
+        cleaned_data = self.clean()
+        instance = super().save()
```

Comment:
Just a nitpick: you're not passing to super the commit param. Not a problem in this case here, but beware this in the future.

Reply:
Thanks :) I'll delete it